source/snapshot/fetch_worker.go

```diff
@@ -0,0 +1,393 @@
+		f.snapshotEnd = t.SnapshotEnd
+		f.lastRead = t.LastRead
+	}
+
+	return f
+}
+
+// Validate will ensure the config is correctt.
+// * Table and keys exist
+// * Key is a primary key
+func (f *FetchWorker) Validate(ctx context.Context) error {
+	if err := f.conf.Validate(); err != nil {
+		return fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	tx, err := f.db.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to start tx for validation: %w", err)
+	}
+	defer func() {
+		if err := tx.Rollback(ctx); err != nil {
+			sdk.Logger(ctx).Warn().
+				Err(err).
+				Msgf("error on validation tx rollback for %q", f.cursorName)
+		}
+	}()
+
+	if err := f.validateTable(ctx, f.conf.Table, tx); err != nil {
+		return fmt.Errorf("failed to validate table: %w", err)
+	}
+
+	if err := f.validateKey(ctx, f.conf.Table, f.conf.Key, tx); err != nil {
+		return fmt.Errorf("failed to validate key: %w", err)
+	}
+
+	return nil
+}
+
+func (f *FetchWorker) Run(ctx context.Context) error {
+	start := time.Now().UTC()
+
+	tx, err := f.db.BeginTx(ctx, pgx.TxOptions{
+		IsoLevel:   pgx.RepeatableRead,
+		AccessMode: pgx.ReadOnly,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := tx.Rollback(ctx); err != nil {
+			sdk.Logger(ctx).Warn().
+				Err(err).
+				Msgf("error run tx rollback for %q", f.cursorName)
+		}
+	}()
+
+	if err := f.withSnapshot(ctx, tx); err != nil {
+		return err
+	}
+
+	if err := f.updateFetchLimit(ctx, tx); err != nil {
+		return fmt.Errorf("failed to update fetch limit: %w", err)
+	}
+
+	closeCursor, err := f.createCursor(ctx, tx)
+	if err != nil {
+		return fmt.Errorf("fail to create cursor: %w", err)
+	}
+	defer closeCursor()
+
+	var nfetched int64
+
+	for {
+		n, err := f.fetch(ctx, tx)
+		if err != nil {
+			return fmt.Errorf("failed to fetch results: %w", err)
+		}
+
+		if n == 0 { // end of cursor
+			break
+		}
+
+		nfetched += int64(n)
+
+		sdk.Logger(ctx).Info().
+			Int64("rows", nfetched).
+			Str("table", f.conf.Table).
+			Dur("elapsed", time.Since(start)).
+			Msg("fetching rows")
+	}
+
+	sdk.Logger(ctx).Info().
+		Dur("elapsed", time.Since(start)).
+		Str("table", f.conf.Table).
+		Msgf("%q snapshot completed", f.conf.Table)
+
+	return nil
+}
+
+func (f *FetchWorker) createCursor(ctx context.Context, tx pgx.Tx) (func(), error) {
+	cursorSQL := fmt.Sprintf("DECLARE %s CURSOR FOR (SELECT * FROM %s WHERE %s > %d AND %s <= %d ORDER BY %s)",
+		f.cursorName,
+		f.conf.Table,
+		f.conf.Key,
+		f.lastRead,
+		f.conf.Key,
+		f.snapshotEnd,
+		f.conf.Key,
+	)
+
+	if _, err := tx.Exec(ctx, cursorSQL); err != nil {
+		return nil, err
+	}
+
+	return func() {
+		// N.B. The cursor will automatically close when the TX is done.
+		if _, err := tx.Exec(ctx, "CLOSE "+f.cursorName); err != nil {
+			sdk.Logger(ctx).Warn().
+				Err(err).
+				Msgf("unexpected error when closing cursor %q", f.cursorName)
+		}
+	}, nil
+}
+
+func (f *FetchWorker) updateFetchLimit(ctx context.Context, tx pgx.Tx) error {
```

Comment:
Nitpick: I'd either rename this method to `updateSnapshotEnd` or rename the field to `fetchLimit`, so the function name and the field it's updating match.